rfcs/0000-buildpack-yml-to-env-var.md

```diff
@@ -0,0 +1,86 @@
+# Buildpack.yml to Environment Variables
+
+## Proposal
+
+Migrate to using environment variables to do all buildpack configuration and
+get rid of `buildpack.yml`.
+
+## Motivation
+
+There are several reasons for making this switch.
+1. There is already an existing RFC that proposes moving away from
+   `buildpack.yml` as a configuration tool.
+1. Environment variables appears to be the standard for configuration in other
+   buildpack ecosystems such as Google Buildpacks and Heroku as well as the
+   Paketo Java buildpacks. Making this change will align the buildpack with the
+   rest of the buildpack ecosystem.
+1. There is native support to pass environment variables to the buildpack
+   either on a per run basis or by configuration that can be checked into
+   source control, in the form of `project.toml`.
+
+## Implementation
+
+The proposed environment variables for Go Build are as follow:
+
+#### BP_GO_BUILD_FLAGS
+
+```shell
+$BP_GO_BUILD_FLAGS='-buildmode=default -tags=paketo -ldflags="-X main.variable=some-value"'
```

Comment:
I found the original comment that this is inspired from the `GOFLAGS` env var. By that nature, does it also inherit the same limitation where spaces aren't allowed?

> Because the entries are space-separated, flag values must not contain spaces. 

If so `-ldflags="-X main.variable=some-value"` wouldn't work right?

Or will this implementation have special parsing where flag values will be parsed as per unresolved issue https://github.com/golang/go/issues/26849?

Reply:
My thought process was to just treat the value as a string that would be passed in its entirety as one argument during build. I cannot think of a downside of doing this, but I am not the most well versed the go flags so if there is a problem please let me know.

Reply:
Looking at the [current implementation](https://github.com/paketo-buildpacks/go-build/blob/main/go_build_process.go#L76) it would most likely not work. If you pass the value of the env var as a single value it wouldn't be a valid argument.

If you pass it as an environment variable `GOFLAGS` then you'll run into the limitation mentioned (where `-ldflags="-X main.variable=some-value"` wouldn't work).

IMO there are two options:

1. pass the value as `GOFLAGS` as an Environment Variable to the command and call out the limitation based on https://github.com/golang/go/issues/26849 and wait for it to be resolved by the `go` community.
2. use custom parsing to parse that value into independent flags/values where flag values can be escaped via double quotes.

I know @ekcasey has done some more extensive cmd arg parsing (and passthrough) in `go` so she might have some insight.

Reply:
After some discussion and screwing around in a [playground](https://play.golang.org/p/MFMnhYyXel9), I think that I have determined that it is not too taxing to just write a custom parser to do this. I think that the idea is to parse split on every space that is not escaped by either `"` or `'`. I think that this does set a precedent for future buildpacks that would like to do configuration of CLI flags through env vars. I will obviously look into making a more robust parser and I will update the RFC soon to reflect this decision.

Reply:
Here is an [extension](https://github.com/ForestEckhardt/semaphore) on the parser that was in the playground this one is fleshed out and has tests and what not.

Reply:
Also putting this here for context. https://github.com/mattn/go-shellwords

Reply:
https://github.com/GoogleCloudPlatform/buildpacks#go-buildpacks